src/Domain/Services/ExtendRestApi.php

```diff
@@ -0,0 +1,193 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+use Automattic\WooCommerce\Blocks\Domain\Package;
+use Automattic\WooCommerce\Blocks\StoreApi\Schemas\CartItemSchema;
+use Throwable;
+use Exception;
+
+/**
+ * Service class to provide utility functions to extend REST API.
+ */
+class ExtendRestApi {
+	/**
+	 * Holds the Package instance
+	 *
+	 * @var Package
+	 */
+	private $package;
+
+	/**
+	 * Constructor
+	 *
+	 * @param Package $package An instance of the package class.
+	 */
+	public function __construct( Package $package ) {
+		$this->package = $package;
+	}
+	/**
+	 * Valid endpoints to extend
+	 *
+	 * @var array
+	 */
+	private $endpoints = [ CartItemSchema::IDENTIFIER ];
+
+	/**
+	 * Data to be extended
+	 *
+	 * @var array
+	 */
+	private $extend_data = [];
+
+	/**
+	 * An endpoint that validates registration method call
+	 *
+	 * @param string   $endpoint The endpoint to extend.
+	 * @param string   $namespace Plugin namespace.
+	 * @param callable $schema_callback Callback executed to add schema data.
+	 * @param callable $data_callback Callback executed to add endpoint data.
+	 *
+	 * @throws Exception On failure to register.
+	 * @return boolean True on success.
+	 */
+	public function register_endpoint_data( $endpoint, $namespace, $schema_callback, $data_callback ) {
+		if ( ! is_string( $namespace ) ) {
+			$this->throw_exception(
+				__( 'You must provide a plugin namespace when extending a Store REST endpoint.', 'woo-gutenberg-products-block' )
+			);
+		}
+
+		if ( ! is_string( $endpoint ) || ! in_array( $endpoint, $this->endpoints, true ) ) {
+			$this->throw_exception(
+				/* translators: 1: a list of endpoints. 2: endpoint provided by user. */
+				sprintf( __( 'You must provide a valid Store REST endpoint to extend, valid endpoints are: %1$s. You provided %2$s.', 'woo-gutenberg-products-block' ), implode( ', ', $this->endpoints ), $endpoint )
+			);
+		}
+
+		if ( ! is_callable( $schema_callback ) ) {
+			$this->throw_exception(
+				__( '$schema_callback must be a callable function.', 'woo-gutenberg-products-block' )
+			);
+		}
+
+		if ( ! is_callable( $data_callback ) ) {
+			$this->throw_exception(
+				__( '$data_callback must be a callable function.', 'woo-gutenberg-products-block' )
+			);
+		}
+
+		$this->extend_data[ $endpoint ][ $namespace ] = [
+			'schema_callback' => $schema_callback,
+			'data_callback'   => $data_callback,
+		];
+
+		return true;
+	}
+
+	/**
+	 * Returns the registered endpoint data
+	 *
+	 * @param string $endpoint    A valid identifier.
+	 * @param array  $passed_args Passed arguments from the Schema class.
+	 * @return array Returns an array with registered endpoint data.
+	 * @throws Exception If a registered callback throws an error, or silently logs it.
+	 */
+	public function get_endpoint_data( $endpoint, array $passed_args = [] ) {
+		$registered_data = (object) [];
```

Comment:
This produces a fatal error in testing.

This should be left as an array until returned because on line 115 you are adding to the array.

So swap this to `return (object) $registered_data;` and update the docblock, assuming objects are expected where this method is called.